api/src/main/java/org/apache/livy/LivyClientBuilder.java

```
@@ -118,24 +121,21 @@ public LivyClient build() {
     }
 
     LivyClient client = null;
-    ServiceLoader<LivyClientFactory> loader = ServiceLoader.load(LivyClientFactory.class,
-      classLoader());
-    if (!loader.iterator().hasNext()) {
-      throw new IllegalStateException("No LivyClientFactory implementation was found.");
```

now we are missing this check, aren't we?

---

I think it is ok to not perform this check. If the iterator is empty, then after the for loop below, client will be null, and an exception (in this case, an IllegalArgumentException) will be thrown. So there will still be an error. I am trying to minimize the code that gets called in the synchronized block. I don't feel this check adds a lot of value, because this IllegalStateException will only happen if the services file is not in the build, which would be a build issue. If you feel very strongly about having this check, I can put it back in.
The other reason I took it out was that ideally this check should be done just once. The normal way to do that would be in a static initializer, but you normally should not throw an exception in those (you can throw a RuntimeException, but I think that is still bad form).

---

You could also simply set a flag in the new for loop that iterates over factories. Immediately after the loop, if the flag is not set, you can throw an exception ("No LivyClientFactory implementation was found."). That *might* be cheaper than first creating an iterator and calling hasNext... but I don't know that.

---

I think we can also do this only once when the `ServiceLoader` is created. But I do see value in keeping the behavior consistent so that a user who faced this issue once, knows what to do facing it again.

---

I'll restore a check for whether any LivyClientFactory implementation was found.
@bersprockets thanks for the suggestion.